### ISSUE TITLE
OwnCloud 8.x problem with patchworks 1.2.2 on rhel/centos 7.x

### ISSUE THREAD

**Issue by paoloB132** [tchwork/utf8#44]
See issue https://github.com/owncloud/core/issues/16654

It seems that some problem are in place with patchworks utf-8 library used by owcloud 8, when on rhel/centos 7.x

They call on file owncloud/apps/user_ldap/user_ldap.php

//find out dn of the user name
```
		$attrs = array($this->access->connection->ldapUserDisplayName, 'dn',
			'uid', 'samaccountname');
		$filter = \OCP\Util::mb_str_replace(
			'%uid', $uid, $this->access->connection->ldapLoginFilter, 'utf-8');

to define correct dn. But $uid is not correctly returned.
\OCP\Util::mb_str_replace calls:

/**
	 * Replace all occurrences of the search string with the replacement string
	 *
	 * @param string $search The value being searched for, otherwise known as the needle. String.
	 * @param string $replace The replacement string.
	 * @param string $subject The string or array being searched and replaced on, otherwise known as the haystack.
	 * @param string $encoding The encoding parameter is the character encoding. Defaults to UTF-8
	 * @param int $count If passed, this will be set to the number of replacements performed.
	 * @return string
	 */
	public static function mb_str_replace($search, $replace, $subject, $encoding = 'UTF-8', &$count = null) {
		return(\OC_Helper::mb_str_replace($search, $replace, $subject, $encoding, $count));
	}

Where \OC_Helper::mb_str_replace correspond to:

	/**
	 * Replace all occurrences of the search string with the replacement string
	 *
	 * @param string $search The value being searched for, otherwise known as the needle.
	 * @param string $replace The replacement
	 * @param string $subject The string or array being searched and replaced on, otherwise known as the haystack.
	 * @param string $encoding The encoding parameter is the character encoding. Defaults to UTF-8
	 * @param int $count If passed, this will be set to the number of replacements performed.
	 * @return string
	 *
	 */
	public static function mb_str_replace($search, $replace, $subject, $encoding = 'UTF-8', &$count = null) {
		$offset = -1;
		$length = mb_strlen($search, $encoding);
				
				
		while (($i = mb_strrpos($subject, $search, $offset, $encoding)) !== false) {
			$subject = OC_Helper::mb_substr_replace($subject, $replace, $i, $length);
			$offset = $i - mb_strlen($subject, $encoding);
			$count++;
		}
		return $subject;
	}
```
Where OC_Helper::mb_substr_replace is:
```
/**
	 * replaces a copy of string delimited by the start and (optionally) length parameters with the string given in replacement.
	 *
	 * @param string $string
	 * @param string $replacement The replacement string.
	 * @param int $start If start is positive, the replacing will begin at the start'th offset into string. If start is negative, the replacing will begin at the start'th character from the end of string.
	 * @param int $length Length of the part to be replaced
	 * @param string $encoding The encoding parameter is the character encoding. Defaults to UTF-8
	 * @internal param string $input The input string. .Opposite to the PHP build-in function does not accept an array.
	 * @return string
	 */
	public static function mb_substr_replace($string, $replacement, $start, $length = null, $encoding = 'UTF-8') {
		$start = intval($start);
		$length = intval($length);
		$string = mb_substr($string, 0, $start, $encoding) .
			$replacement .
			mb_substr($string, $start + $length, mb_strlen($string, 'UTF-8') - $start, $encoding);

		return $string;
	}
```

And finally mb_substr is part of /utf8/class/Patchwork/PHP/Shim/Mbstring.php  Patchwork library.

Returned $string is empty.

So it seems that Patchworks doesn't work as expected. Have you tested Patchwork on rhel/centos 7 ? Are you aware of possible troubles ?

Regards,
Paolo.

**Comment by nicolas-grekas** [comment]
So, RHEL doesn't have the mbstring extension and they rely on the patchwork/utf-8 fallback? They should enable mbstring.
An other question: why do you need mb_(sub)str_replace in the first place? String substitution is encoding safe, there is no need for coping with multi bytes things here, just use (sub)str_replace directly.

That said, it does not prevent me from fixing a bug in the fallback :)
I'll look at this and tell you if I manage to reproduce the issue.

**Comment by paoloB132** [comment]
Hi Nicolas,
 
> On 19/giu/2015, at 13:17, Nicolas Grekas <notifications@github.com> wrote:
> 
> So, RHEL doesn't have the mbstring extension and they rely on the patchwork/utf-8 fallback? They should enable mbstring.
Really php-mbstring is available

php-mbstring.x86_64              5.4.16-23.el7_0.3   rhel-7-server-optional-rpms

> An other question: why do you need mb_(sub)str_replace in the first place? String substitution is encoding safe, there is no need for coping with multi bytes things here, just use (sub)str_replace directly.
> 
I simply found out the bug as user,…OC developers are responsible to use patchwork libs on OC 8…

> That said, it does not prevent me from fixing a bug in the fallback :)
> I'll look at this and tell you if I manage to reproduce the issue.
> 
Great…thanks.

> —
> Reply to this email directly or view it on GitHub <https://github.com/tchwork/utf8/issues/44#issuecomment-113480089>.
> 

------------------------------------------------------------------------------------------------
Paolo Barbato

Consorzio RFX
corso Stati Uniti,4                                  
35127 Padova - Italy                     	                 
Network Administrator 
phone: +39 049 8295097 fax: +39 049 8700718
------------------------------------------------------------------------------------------------

**Comment by nicolas-grekas** [comment]
So, if the mbstring fallback of patchwork/utf8 is not used, then this issue has nothing to do with patchwork/utf8... closing

**Comment by paoloB132** [comment]
Nicolas...anyway the bug seems there !

**Comment by nicolas-grekas** [comment]
Sure, but not in patchwork/utf-8, since you just said mbstring is enabled, which means patchwork's code is not called at all.

**Comment by paoloB132** [comment]
sure...thanks again!

**Comment by nicolas-grekas** [comment]
So, reading comments on the original thread, mbstring was not enabled, which means we ay have aa bug here. Reopening.

**Comment by nicolas-grekas** [comment]
Fixed in #47 thank you @paoloB132
